android/src/main/java/com/mattermost/networkclient/APIClientModule.kt

```
@@ -200,6 +201,73 @@ internal class APIClientModule(reactContext: ReactApplicationContext) : ReactCon
         request("DELETE", baseUrl, endpoint, options, promise)
     }
 
+    @ReactMethod
+    fun download(baseUrl: String, endpoint: String, filePath: String, taskId: String, options: ReadableMap?, promise: Promise) {
+        var url: HttpUrl
+        try {
+            url = baseUrl.toHttpUrl()
+        } catch (error: IllegalArgumentException) {
+            return promise.reject(error)
+        }
+
+        val client = clients[url]!!
+        val downloadCall = client.buildDownloadCall(endpoint, taskId, options)
+        calls[taskId] = downloadCall
+
+        try {
+            downloadCall.enqueue(object : okhttp3.Callback {
+                override fun onFailure(call: Call, e: IOException) {
+                    calls.remove(taskId)
+                    promise.reject(e)
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                    var inputStream: InputStream? = null
+                    var outputStream: FileOutputStream? = null
+                    try {
+                        val responseBody = response.body
+                        if (responseBody != null) {
+                            val f = File(filePath)
+                            val parent = f.parentFile
+                            if (parent != null && !parent.exists() && !parent.mkdirs()) {
+                                promise.reject(Error("Couldn't create dir: " + parent.path))
```

have this fail early